### PR TITLE
Return meaningful values for MessageTask getMethodName function[API-2037]

### PR DESCRIPTION
**Pull request by ihsandemir**
Return meaningful values for MessageTask `getMethodName` function. Added non-null method names to some of the message tasks which were missing the names and can be used in security interceptor since they are related to the user method calls.

The ones which return null for the `getMethodName` function are the internally used or not directly related to the  user API calls.

Also, removed the default null returning implementations in some of the base message tasks.

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/6252

### CODE REVIEW

**Comment by ihsandemir** [comment]
The following is the list of message tasks which will continue returning null after this PR is merged. 

```
Targets
    Occurrences of '    public String getMethodName\(\) \{.*
    .*return null;
    \}' in Directory /Users/ihsan/Desktop/work/src/hazelcast
Found Occurrences in Directory /Users/ihsan/Desktop/work/src/hazelcast  (34 usages found)
    Production  (34 usages found)
        Unclassified  (34 usages found)
            hazelcast  (34 usages found)
                com.hazelcast.client.impl.protocol.task  (12 usages found)
                    AddBackupListenerMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            89 public String getMethodName() {
                    AddClusterViewListenerMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            74 public String getMethodName() {
                    AuthenticationBaseMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            256 public String getMethodName() {
                    ClientStatisticsMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            68 public String getMethodName() {
                    CreateProxiesMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            98 public String getMethodName() {
                    CreateProxyMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            98 public String getMethodName() {
                    DeployClassesMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            93 public String getMethodName() {
                    DestroyProxyMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            104 public String getMethodName() {
                    ExperimentalTpcAuthenticationMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            59 public String getMethodName() {
                    NoSuchMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            64 public String getMethodName() {
                    PingMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            68 public String getMethodName() {
                    TriggerPartitionAssignmentMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            67 public String getMethodName() {
                com.hazelcast.client.impl.protocol.task.cache  (6 usages found)
                    CacheAddNearCacheInvalidationListenerTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            105 public String getMethodName() {
                    CacheDestroyMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            78 public String getMethodName() {
                    CacheFetchNearCacheInvalidationMetadataTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            78 public String getMethodName() {
                    CacheGetConfigMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            62 public String getMethodName() {
                    CacheManagementConfigMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            79 public String getMethodName() {
                    CacheRemoveInvalidationListenerMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            79 public String getMethodName() {
                com.hazelcast.client.impl.protocol.task.crdt.pncounter  (1 usage found)
                    PNCounterGetConfiguredReplicaCountMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            73 public String getMethodName() {
                com.hazelcast.client.impl.protocol.task.map  (5 usages found)
                    MapFetchNearCacheInvalidationMetadataTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            78 public String getMethodName() {
                    MapMadePublishableMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            68 public String getMethodName() {
                    MapPublisherCreateMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            148 public String getMethodName() {
                    MapPublisherCreateWithValueMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            151 public String getMethodName() {
                    MapSetReadCursorMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            67 public String getMethodName() {
                com.hazelcast.client.impl.protocol.task.schema  (3 usages found)
                    FetchSchemaMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            68 public String getMethodName() {
                    SendAllSchemasMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            70 public String getMethodName() {
                    SendSchemaMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            71 public String getMethodName() {
                com.hazelcast.client.impl.protocol.task.transaction  (5 usages found)
                    XAClearRemoteTransactionMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            82 public String getMethodName() {
                    XACollectTransactionsMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            98 public String getMethodName() {
                    XAFinalizeTransactionMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            66 public String getMethodName() {
                    XATransactionCreateMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            70 public String getMethodName() {
                    XATransactionPrepareMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            75 public String getMethodName() {
                com.hazelcast.cp.internal.datastructures.semaphore.client  (1 usage found)
                    GetSemaphoreTypeMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            71 public String getMethodName() {
                com.hazelcast.cp.internal.session.client  (1 usage found)
                    AbstractSessionMessageTask.java  (1 usage found)
                        getMethodName()  (1 usage found)
                            62 public String getMethodName() {
```

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheCreateConfigMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheCreateConfigMessageTask.java:93]: is there a user method as createConfig for ICache? I think this should be createCacheConfigOnAllMembers

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheListenerRegistrationMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheListenerRegistrationMessageTask.java:82]: I think this message task is used by many user methods. However, I think registerListener is a better name because many of the registration methods start with registerListener

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToAddressMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToAddressMessageTask.java:98]: I think the user facing method name is executeOnMember

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToPartitionMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToPartitionMessageTask.java:90]: The user methods using this message task are submit and submitToKeyOwner. I think we should use one of these. WDYT?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapDestroyCacheMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapDestroyCacheMessageTask.java:89]: There is no QueryCache.destroy() that the user can use to call this message task. Or they can call IMap.destroy as well. I think destroyQueryCache is a better name.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCommitMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCommitMessageTask.java:77]: The method name should be commitTransaction IMO. See TransactionContext.commitTransaction

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCreateMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCreateMessageTask.java:89]: Similarly this should be beginTransaction

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionRollbackMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionRollbackMessageTask.java:77]: Similarly this should be rollbackTransaction

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionCommitMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionCommitMessageTask.java:80]: The method name for this is just commit. See XAResource.commit.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionRollbackMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionRollbackMessageTask.java:77]: The method name for this is just rollback. See XAResource.rollback.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheListenerRegistrationMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheListenerRegistrationMessageTask.java:82]: for the ones which do not directly correspond to one to one user API (this can also be called on the background when client reconnects, etc.) I tried to use the same name as the internal message task name. I tried to do as suggested in [this](https://hazelcast.atlassian.net/browse/SUP-225?focusedCommentId=77717) comment. @kwart Can you also comment what is a proper name for such message tasks based on how the user can use it int he interceptor. I am open to suggestions.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToAddressMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToAddressMessageTask.java:98]: This task is called using multiple different user APIs, these are :
```
    public <T> Future<T> submit(@Nonnull Callable<T> task,
                                @Nonnull MemberSelector memberSelector);

    public <T> Map<Member, Future<T>> submitToMembers(@Nonnull Callable<T> task,
                                                      @Nonnull Collection<Member> members) ;

    public <T> Map<Member, Future<T>> submitToAllMembers(@Nonnull Callable<T> task) ;

....
....
```
Hence, for such cases we use a common name e.g. the message task name as suggested at this comment.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToPartitionMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToPartitionMessageTask.java:90]: See above [comment](https://github.com/hazelcast/hazelcast/pull/25020#discussion_r1265022035).

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapDestroyCacheMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapDestroyCacheMessageTask.java:89]: It is called through [this](https://github.com/hazelcast/hazelcast/blob/v5.3.0/hazelcast/src/main/java/com/hazelcast/map/QueryCache.java#L226) API. It is NOT IMap method.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCommitMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCommitMessageTask.java:77]: changed.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionRollbackMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionRollbackMessageTask.java:77]: changed.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionRollbackMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionRollbackMessageTask.java:77]: changed.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCreateMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/TransactionCreateMessageTask.java:89]: changed.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionCommitMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XATransactionCommitMessageTask.java:80]: changed.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToAddressMessageTask.java
Comment (srknzl) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/executorservice/ExecutorServiceSubmitToAddressMessageTask.java:98]: executeOnMember and executeOnMembers use this message task as well. Anyway, to choose a common name, lets use what you suggest.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapDestroyCacheMessageTask.java
Comment (srknzl) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapDestroyCacheMessageTask.java:89]: right, but it can also be called through IMap.destroy. Lets follow task name.

**Review by srknzl — APPROVED** [review]
lgtm except I think we should make sure we document these if the user will need to use these strings. according to https://docs.hazelcast.com/imdg/4.2/security/security-interceptor#latest-banner, the user needs to know the exact method name.  

If they need to know the exact strings, we should list these strings in our documentation, together with the methods each string corresponds to. e.g "destroyCache" -> "QueryCache.destroy" and "IMap.destroy"

Another question: Can we use the same strings (e.g using "create" for two different message tasks) in two different getMethodName's? What happens in that case ? If it is not allowed, do we validate if we are not duplicating any string in getMethodName?

**Comment by ihsandemir** [comment]
> 

@srknzl  The unique one is service name + method name, hence, it is ok to have the same method name e.g. using "create" for two different message tasks since their service names will differ.

**Comment by ihsandemir** [comment]
> Could you place the names in a separate constant class to make it a public API?
> 
> I think we need to cover more classes - [the comment](https://hazelcast.atlassian.net/browse/SUP-225?focusedCommentId=77464) mentions `CreateProxyMessageTask` and `DestroyProxyMessageTask`.
> 
> Also these ones from your list caught my eye:
> 
> ```
>                     DeployClassesMessageTask.java  (1 usage found)
>                     CacheDestroyMessageTask.java  (1 usage found)
> ```
> 
> We could also consider the ones with `*Config*` in the name.

I will put them in a new constants file `com.hazelcast.security.SecurityInterceptorConstants` as requested.

What about the `objectType` (service names)? Should we also have them among the constants? The user needs them.

Can the user do security check on proxy creation which happens for each data structure?

**Comment by ihsandemir** [comment]
> Could you place the names in a separate constant class to make it a public API?
> 
> I think we need to cover more classes - [the comment](https://hazelcast.atlassian.net/browse/SUP-225?focusedCommentId=77464) mentions `CreateProxyMessageTask` and `DestroyProxyMessageTask`.
> 
> Also these ones from your list caught my eye:
> 
> ```
>                     DeployClassesMessageTask.java  (1 usage found)
>                     CacheDestroyMessageTask.java  (1 usage found)
> ```
> 
> We could also consider the ones with `*Config*` in the name.

@kwart @srknzl  What about using enums instead of free String in the parameter.
```
diff --git a/hazelcast/src/main/java/com/hazelcast/security/SecurityInterceptor.java b/hazelcast/src/main/java/com/hazelcast/security/SecurityInterceptor.java
index 65a583a409..bf6cca5198 100644
--- a/hazelcast/src/main/java/com/hazelcast/security/SecurityInterceptor.java
+++ b/hazelcast/src/main/java/com/hazelcast/security/SecurityInterceptor.java
@@ -23,6 +23,20 @@ import java.security.AccessControlException;
  */
 public interface SecurityInterceptor {
+    public enum ObjectName {
+        PUT(“put”);
+
+        private final String methodName;
+
+        ObjectName(String methodName) {
+            this.methodName = methodName;
+        }
+
+        public String getMethodName() {
+            return methodName;
+        }
+    }
+
     /**
      *
      * @param credentials
@@ -43,5 +57,5 @@ public interface SecurityInterceptor {
      * @param methodName
      * @param parameters
      */
-    void after(Credentials credentials, String objectType, String objectName, String methodName, Parameters parameters);
+    void after(Credentials credentials, String objectType, ObjectName objectName, String methodName, Parameters parameters);
 }
```
Is there any specific reason to make the parameter as free String?
We deprecate the old one and put the new one in place. what do you think?

**Comment by kwart** [comment]
As discussed internally, the enums (compared to free Strings) can lead to `NoSuchFieldError` thrown. I.e.
* for backward compatibility you wouldn't be allowed to remove a value/constant from enum in newer Hazelcast versions;
* interceptor developed for one version might not work with older Hazelcast versions.

**Review by srknzl — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheIterateMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheIterateMessageTask.java:81]: Should not this be SecurityInterceptorConstants.ITERATE? 🤔

**Comment by ihsandemir** [comment]
I am not putting any security checks on the compact schema management tasks at `package com.hazelcast.client.impl.protocol.task.schema;`. These are for internal usage.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheIterateMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheIterateMessageTask.java:81]: They actually correspond to the hasNext and next methods of the iterator. I am changing it back to iterator as suggested.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheRemovePartitionLostListenerMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheRemovePartitionLostListenerMessageTask.java:79]: Yes, it is intentional. There is the `getServiceName` method as well at the Security interceptor which identifies if it is cache service or not, hence no need for the work `Cache` in the method name and I removed it.

**Comment by ihsandemir** [comment]
@kwart @srknzl I updated the Pr based on comments from Josef. Can you guys review the latest?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XACollectTransactionsMessageTask.java
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XACollectTransactionsMessageTask.java:100]: collect transactions and recover? What is the relation ?

**Review by srknzl — APPROVED** [review]
approving with one comment above

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XACollectTransactionsMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/transaction/XACollectTransactionsMessageTask.java:100]: When I trace this task and try to find the associated user API, I found that the user API is `XAResource.recover` [API](https://docs.oracle.com/javase/8/docs/api/javax/transaction/xa/XAResource.html#recover-int-) and hence I named it it this way to relate to the user API.

**Comment by kwart** [comment]
run-ee-tests

**Comment by kwart** [comment]
run-ee-nightly-tests

**Review by kwart — CHANGES_REQUESTED** [review]
Could you also send a fix for the 2 EE tests that fail with the new names?

If we proceed with the rename (i.e. change from a not-null value to another one), then these instances should be listed in the Release Notes as breaking changes. We should make doc and support teams aware of it.

Overall, I like the changes. The majority of my comments are just process ones :)

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CachePutAllMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CachePutAllMessageTask.java:78]: This is to make the interceptor test at [here](https://github.com/hazelcast/hazelcast-enterprise/blob/master/hazelcast-enterprise/src/test/java/com/hazelcast/client/security/CacheSecurityInterceptorTest.java#L105) work properly and it makes sense since the parameter of the `ICache.putAll` method receives a map. Yes, it is one more map creation but it makes the API more user friendly. Let me know if it is OK or I am happy to revert it if we agree on returning the `parameters.entries`
```     
public java.util.List<java.util.Map.Entry<com.hazelcast.internal.serialization.Data, com.hazelcast.internal.serialization.Data>> entries;
```

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/scheduledexecutor/ScheduledExecutorGetAllScheduledMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/scheduledexecutor/ScheduledExecutorGetAllScheduledMessageTask.java:92]: Yes, it is intended to match the actual user API at [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/IScheduledExecutorService.java#L451)

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/scheduledexecutor/ScheduledExecutorSubmitToPartitionMessageTask.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/scheduledexecutor/ScheduledExecutorSubmitToPartitionMessageTask.java:88]: The `ExecutorService*`, `DurableExecutor*`  are not really directly related to the `IScheduledExecutorService`, hence, this is different to match what `IScheduledExecutorService` uses. Can you also check and we can discuss again if you like.

**Review by kwart — APPROVED** [review]
Thanks for the enhancements, Ihsan. :+1:

**Comment by ihsandemir** [comment]
run-lab-run

**Pull request merged**
